### PR TITLE
Add API usage telemetry to gh-helper commands

### PR DESCRIPTION
**Pull request by apstndb**
## Summary
- Add `--api-usage` to include command-local REST/GraphQL usage telemetry in structured output or stderr summaries.
- Inject GraphQL `rateLimit` fields into query operations so `cost`, `nodeCount`, `limit`, `remaining`, `used`, `resetAt`, and `resetInSeconds` come from API responses when available.
- Track response header rate-limit budgets for REST and GraphQL calls, and warn on stderr when usage crosses the configurable rate-limit threshold.
- Address Copilot/Gemini review feedback by omitting empty rate-limit objects, using Cobra stderr streams, and marking structured telemetry only after successful encoding.

Fixes #57.

## Validation
- `make check`
- `make build`
- `./bin/gh-helper reviews bot-status 56 --json --api-usage --jq '.apiUsage'`
- `./bin/gh-helper reviews bot-status 56 --json --rate-limit-warning-threshold 0.01 --jq '.botStatus.pr'`
- `./bin/gh-helper reviews bot-status 63 --json --api-usage --jq '.apiUsage'`
- OpenCode diff review on PR #63: no material issues found

### CODE REVIEW

File: gh-helper/api_usage.go [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+	}
+}
+
+func apiUsageEnabledForCommand(cmd *cobra.Command) bool {
+	if cmd == nil {
+		return false
+	}
+	enabled, err := cmd.Root().PersistentFlags().GetBool("api-usage")
+	return err == nil && enabled
+}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go]: Addressed in 1a7873e.

Fixed in commit 1a7873e.

File: gh-helper/graphql_rate_limit.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+			s.pos++
+			return
+		}
+		s.pos++
+	}
+}
+
+func (s *graphQLScanner) skipBlockString() {
+	s.pos += 3
+	if end := bytes.Index([]byte(s.input[s.pos:]), []byte(`"""`)); end >= 0 {
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/graphql_rate_limit.go]: Addressed in 1a7873e.

Fixed in commit 1a7873e.

File: gh-helper/api_usage.go [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+	"time"
+
+	"github.com/spf13/cobra"
+)
+
+type APIUsageReport struct {
+	Backend   string            `json:"backend"`
+	Observed  APIUsageObserved  `json:"observed"`
+	RateLimit APIUsageRateLimit `json:"rateLimit,omitempty"`
+	Warnings  []string          `json:"warnings,omitempty"`
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go]: Addressed in 1a7873e.

Fixed in commit 1a7873e.

File: gh-helper/api_usage.go [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+	}
+	duration = duration.Round(time.Second)
+	if duration < time.Minute {
+		return duration.String()
+	}
+	duration = duration.Round(time.Minute)
+	if duration < time.Hour {
+		return duration.String()
+	}
+	duration = duration.Round(time.Minute)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go]: Addressed in 1a7873e.

Fixed in commit 1a7873e.

**Comment by apstndb** [comment]
OpenCode diff review completed for `1a7873e`.

Result: no material issues found.

Non-blocking observations from OpenCode were limited to theoretical scanner edge cases and optional extra tests; the concrete test and implementation gaps raised by Copilot/Gemini were addressed in `1a7873e`.

File: gh-helper/api_usage.go [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+
+	if dataMap, ok := data.(map[string]interface{}); ok {
+		dataMap["apiUsage"] = report
+		return dataMap, true
+	}
+
+	return map[string]interface{}{
+		"data":     data,
+		"apiUsage": report,
+	}, true
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go]: Addressed in dd122e3: --api-usage structured output now always wraps payloads as data plus apiUsage, so map and struct outputs have the same root shape.

Fixed in commit dd122e3.

File: gh-helper/api_usage.go [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+			r.observed.GraphQLCost += used - *previousUsed
+			r.observed.GraphQLCostEstimatedRequests++
+		} else {
+			r.observed.GraphQLCostUnavailableRequests++
+		}
+		r.lastGraphQLUsed = apiUsageIntPtr(used)
+		return
+	}
+	r.observed.GraphQLCostUnavailableRequests++
+}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go]: Addressed in dd122e3: warning-only tracking now uses headers only and skips GraphQL response body parsing unless --api-usage or --api-usage-trace needs request-level telemetry.

Fixed in commit dd122e3.

File: gh-helper/api_usage.go [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+		resource,
+		usedRatio*100,
+		rateLimit.Used,
+		rateLimit.Limit,
+		rateLimit.Remaining,
+		resetAt,
+		resetIn,
+	)
+	r.warningWritten = true
+}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go:422]: Addressed in dd122e3: high-watermark warnings are now tracked per resource, so core and graphql can each emit their own warning once per command.

Fixed in commit dd122e3.

File: gh-helper/graphql_rate_limit.go [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+	scanner := graphQLScanner{input: query, pos: start}
+	for {
+		token, pos, ok := scanner.nextToken()
+		if !ok {
+			return -1
+		}
+		if token == "{" {
+			return pos
+		}
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/graphql_rate_limit.go:145]: Addressed in 8f9fa9b: findNextGraphQLSelectionBrace now tracks parentheses and skips braces inside argument lists when locating the root selection set.

Fixed in commit 8f9fa9b.

File: gh-helper/graphql_rate_limit.go [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+			} else {
+				s.skipString()
+			}
+			continue
+		}
+		if r == '{' || r == '}' {
+			pos := s.pos
+			s.pos++
+			return string(r), pos, true
+		}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/graphql_rate_limit.go]: Addressed in 8f9fa9b: the GraphQL scanner now tokenizes parentheses so selection-set detection can handle variable and directive argument lists correctly.

Fixed in commit 8f9fa9b.

File: gh-helper/api_usage.go [hunk 10]
```diff
@@ -0,0 +1,606 @@
...
+import (
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go:13]: Addressed in 8f9fa9b: added bytes import for the rateLimit alias pre-check before response unmarshalling.

Fixed in commit 8f9fa9b.

File: gh-helper/api_usage.go [hunk 11]
```diff
@@ -0,0 +1,606 @@
...
+	r.observed.GraphQLRequests++
+	if rateLimit, ok := graphQLRateLimitFromResponse(body); ok {
+		r.observed.GraphQLCost += rateLimit.Cost
+		r.observed.GraphQLCostKnownRequests++
+		r.observed.GraphQLNodeCount += rateLimit.NodeCount
+		r.updateAfterFromGraphQLRateLimit(rateLimit)
+		r.lastGraphQLUsed = apiUsageIntPtr(rateLimit.Used)
+		r.writeGraphQLTraceLocked(trace, "response", &rateLimit, nil)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go]: Addressed in 8f9fa9b: RecordGraphQLResponse now checks for the telemetry alias before JSON unmarshalling, so responses without injected rateLimit avoid the extra parse.

Fixed in commit 8f9fa9b.

File: gh-helper/api_usage.go [hunk 12]
```diff
@@ -0,0 +1,606 @@
...
+		writer,
+		"warning: GitHub %s API rate limit is %.0f%% used (%d/%d used, %d remaining, resets at %s%s)\n",
+		resource,
+		usedRatio*100,
+		rateLimit.Used,
+		rateLimit.Limit,
+		rateLimit.Remaining,
+		resetAt,
+		resetIn,
+	)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go]: Addressed in 8f9fa9b: rate-limit threshold warnings are now appended to structured apiUsage.warnings when --api-usage is enabled, while still printing to stderr.

Fixed in commit 8f9fa9b.

File: gh-helper/api_usage.go [hunk 13]
```diff
@@ -0,0 +1,612 @@
...
+		if rateLimit, ok := graphQLRateLimitFromResponse(body); ok {
+			r.observed.GraphQLCost += rateLimit.Cost
+			r.observed.GraphQLCostKnownRequests++
+			r.observed.GraphQLNodeCount += rateLimit.NodeCount
+			r.updateAfterFromGraphQLRateLimit(rateLimit)
+			r.lastGraphQLUsed = apiUsageIntPtr(rateLimit.Used)
+			r.writeGraphQLTraceLocked(trace, "response", &rateLimit, nil)
+			return
+		}
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go:228]: Thanks. Keeping the current RawMessage-based extraction for now: RecordGraphQLResponse already pre-checks the telemetry alias before unmarshalling, and the remaining parse only happens when --api-usage or --api-usage-trace needs telemetry. A streaming extractor would add complexity without a demonstrated large-response bottleneck in this CLI path.

File: gh-helper/graphql_rate_limit.go [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+	}
+
+	closeBrace := findMatchingGraphQLBrace(query, openBrace)
+	if closeBrace < 0 {
+		return query, false
+	}
+
+	injected := "\n  " + graphQLRateLimitAlias + ": rateLimit {\n    cost\n    limit\n    nodeCount\n    remaining\n    used\n    resetAt\n  }\n"
+	return query[:closeBrace] + injected + query[closeBrace:], true
+}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/graphql_rate_limit.go:54]: Thanks. I am not moving queries to embedded .graphql files in this PR because the changed code is the generic injector/scanner, not a large static query body. Externalizing this repository's larger GraphQL queries would be a broader refactor and should be handled separately from the API-usage telemetry change.

File: gh-helper/api_usage.go [hunk 15]
```diff
@@ -0,0 +1,612 @@
...
+	report.Used = apiUsageIntPtr(after.Used)
+	if after.Limit != 0 {
+		report.Limit = apiUsageIntPtr(after.Limit)
+	}
+	if after.Reset != 0 {
+		resetAt := time.Unix(after.Reset, 0).UTC()
+		report.ResetAt = resetAt.Format(time.RFC3339)
+		report.ResetInSeconds = apiUsageInt64Ptr(maxDurationSeconds(time.Until(resetAt)))
+	}
+	return report
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/api_usage.go:577]: Addressed in 2b85db9: rate-limit snapshots now track per-field presence, omit unseen used/remaining/reset fields from reports and traces, derive used from limit-remaining when possible, and only emit threshold warnings when enough data was observed.

Fixed in commit 2b85db9.
